File: src/main/java/nl/tue/win/lpg/Edge.java
package nl.tue.win.lpg;

import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Set;

public class Edge extends HashMap<String, Object> {

    private final @NotNull Node source;

    private final @NotNull Node target;

    private final @NotNull Set<String> labels;

    public Edge(@NotNull Node source, @NotNull Node target) {
        this.source = source;
        this.target = target;
        this.labels = Set.of();
    }

    public Edge(@NotNull Node source, @NotNull Node target, String ... labels) {
        this.source = source;
        this.target = target;
        this.labels = Set.of(labels);
    }

    public @NotNull Node getSource() {
        return source;
    }

    public @NotNull Node getTarget() {
        return target;
    }

    public @NotNull Set<String> getLabels() {
        return labels;
    }

    @Override
    public boolean equals(Object that) {
        if (this == that) return true;
        if (that == null || getClass() != that.getClass()) return false;
//        if (!super.equals(that)) return false;

        Edge thatEdge = (Edge) that;

        return getSource().equals(thatEdge.getSource())
                && getTarget().equals(thatEdge.getTarget())
                && getLabels().equals(thatEdge.getLabels());
    }

    @Override
    public int hashCode() {
        int result = super.hashCode();
        result = 31 * result + getSource().hashCode();
        result = 31 * result + getTarget().hashCode();
        result = 31 * result + getLabels().hashCode();
        return result;
    }
}

File: src/main/kotlin/nl/tue/win/lpg/Edge.kt
package nl.tue.win.lpg

import nl.tue.win.lpg.encoder.CyJsonCodec
import nl.tue.win.lpg.encoder.GraphCodec

class Edge(val sourceId: String, val targetId: String, val id: String, vararg labels: String = arrayOf()) {

    val labels: Set<String>
    val properties: HashMap<String, Any> = HashMap()

    init {
        this.labels = setOf(*labels)
    }

    operator fun get(property: String): Any? {
        return properties[property]
    }

    operator fun set(property: String, value: Any) {
        properties[property] = value
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Edge

        if (sourceId != other.sourceId) return false
        if (targetId != other.targetId) return false
        if (labels != other.labels) return false
        if (properties != other.properties) return false

        return true
    }

    override fun hashCode(): Int {
        var result = sourceId.hashCode()
        result = 31 * result + targetId.hashCode()
        result = 31 * result + labels.hashCode()
        result = 31 * result + properties.hashCode()
        return result
    }

    override fun toString(): String {
        return toString(CyJsonCodec)
    }

    fun <GraphType, NodesType, EdgesType, NodeType, EdgeType> toString(encoder: GraphCodec<GraphType, NodesType, EdgesType, NodeType, EdgeType>): String {
        return encoder.encodeEdge(this).toString()
    }
}